contracts/Smt.metasol

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+
+library SMT{{hashname}} {
+    // because maxElements = 2**depth, so 256 will overflow
+    uint8 constant MAX_DEPTH = 255;
+    bytes32 constant internal EMPTY_SUBTREE = 0x00;
+
+    struct MerkleTree {
+        uint8 depth;
+        mapping(uint256 => bytes32) emptySubtreeCache;
+        // tree level => index inside level => element hash
+        mapping(uint256 => mapping(uint256 => bytes32)) tree;
+        mapping(uint256 => bytes) elementData;
+    }
+
+    using SMT{{hashname}} for MerkleTree;
+
+    function initialize(MerkleTree storage tree, uint8 _depth) internal {
+        require(_depth > 0, "Depth must be non-zero");
+        tree.emptySubtreeCache[0] = {{hash}}("");
+        // current depth == 0
+        tree._calculateEmptySubtree(_depth);
+        tree.depth = _depth;
+    }
+
+    function totalElements(MerkleTree storage tree) internal view returns (uint256) {
+        // 1<<depth == 2**depth
+        return 1<<tree.depth;
+    }
+
+    function increaseDepth(MerkleTree storage tree, uint8 depthDifference) internal {
+        require(depthDifference > 0, "depthDifference must be non-zero");
+        uint8 oldDepth = tree.depth;
+        require(depthDifference <= MAX_DEPTH - oldDepth, "Overflow protection");
+        uint8 newDepth = oldDepth + depthDifference;
+
+        tree._calculateEmptySubtree(newDepth);
+
+        for (uint8 level = oldDepth+1; level <= newDepth; level += 1) {
+            tree._updateNode(level, 0);
+        }
+        tree.depth = newDepth;
+    }
+
+    function decreaseDepth(MerkleTree storage tree, uint8 depthDifference) internal {
+        require(depthDifference > 0, "depthDifference must be non-zero");
+        uint8 oldDepth = tree.depth;
+        require(depthDifference < oldDepth, "Underflow protection");
+        uint8 newDepth = oldDepth - depthDifference;
+
+        for (uint8 level = newDepth; level < oldDepth; level += 1) {
+            require(tree.tree[level][1] == EMPTY_SUBTREE, "Subtree must be empty");
+            // If we will delete this:
+            // delete tree[level+1][0];
+            // delete emptySubtreeCache[level+1];
+            // refund will be smaller than gas spent :(
+        }
+        tree.depth = newDepth;
+    }
+
+    function getRoot(MerkleTree storage tree) internal view returns (bytes32) {
+        if (tree.tree[tree.depth][0] == EMPTY_SUBTREE) {
+            return tree.emptySubtreeCache[tree.depth];
+        }
+        return tree.tree[tree.depth][0];
+    }
+
+    function _calculateEmptySubtree(MerkleTree storage tree, uint8 toLevel) internal {
```

Comment:
Sorry for the nitpick, but actually keeping the word "hash" (or "root") is quite informative...